crates/ra_hir/src/ty.rs

```diff
@@ -1042,12 +1046,26 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
             },
             Expr::Tuple { exprs } => {
                 let mut ty_vec = Vec::with_capacity(exprs.len());
-                for arg in exprs.iter() {
-                    ty_vec.push(self.infer_expr(*arg, &Expectation::none())?);
+                for expr in exprs.iter() {
+                    ty_vec.push(self.infer_expr(*expr, &Expectation::none())?);
                 }
 
                 Ty::Tuple(Arc::from(ty_vec))
             }
+            Expr::Array { exprs } => {
+                let mut ty = self.insert_type_vars(expected.ty.clone());
```

Comment:
maybe let's call this `elem_ty`, since it should be the element type. So you can't just use `expected.ty`, that's the expected type for the whole slice ;) You need to check whether the expected type is `Slice` or `Array`; if so, you can take the element type from that; otherwise, you can use a new type variable for the element type (`self.new_type_var()`)

Reply:
> You need to check whether the expected type is Slice or Array; if so, you can take the element type from that

How to check whether the expected type is Slice or Array?
I think that there is no information where this processing context.

Reply:
I think I defined wrong ArrayExpr.
ArrayExpr must contains SEMI and Number or Ref referring to usize.
My definitions is not for Array but for Slice, I think now.
Do I have the misconception something?

Reply:
> How to check whether the expected type is Slice or Array?

You have the expected type in `expected.ty` here. You can match on that, basically:
```rust
let elem_ty = match expected.ty {
    Ty::Slice(inner) | Ty::Array(inner) => inner.clone(),
    _ => self.new_type_var()
}
```

Reply:
> I think I defined wrong ArrayExpr.
> ArrayExpr must contains SEMI and Number or Ref referring to usize.
> My definitions is not for Array but for Slice, I think now.
> Do I have the misconception something?

Both `[a, b, c]` and `[a; 99]` are array expressions resulting in an array type, not a slice, and if I'm not mistaken, they're both represented by `ArrayExpr`. You could add support for the second form already here as well, but it'd be fine to ignore that for now and do in in a later PR.

(You get a slice by a coercion from an array, which we don't implement yet, but it will kind of work if we pass down the expected element type like I wrote: for example, in `let x: &[u8] = [1, 2, 3]`, the expected type is a slice of `u8`; then when typing the array expression, we take the expected element type, so `u8`, and use that for the elements. Even though the resulting type of the expression, which is an array of `u8`, doesn't unify with the expected type in the end.)